From: Jiri Kosina <jkosina@suse.cz>
Subject: x86/bugs: spec_ctrl must be cleared from cpu_caps_set when being disabled
References: bsc#1096140
Patch-mainline: Never, SUSE-specific

apply_forced_caps() first removes the force-removed bits, and then adds
those from cpu_caps_set, therefore X86_FEATURE_SPEC_CTRL end up in the
final set even through setup_clear_cpu_cap() has been called for it
(setup_force_cpu_cap() vse. setup_clear_cpu_cap() are not symmetric
unfortunately).

Manually clear the bit from cpu_caps_set when disabling it.

Signed-off-by: Jiri Kosina <jkosina@suse.cz>
---
 arch/x86/kernel/cpu/spec_ctrl.c |    6 ++++++
 1 file changed, 6 insertions(+)

--- a/arch/x86/kernel/cpu/spec_ctrl.c
+++ b/arch/x86/kernel/cpu/spec_ctrl.c
@@ -87,7 +87,13 @@ void __init noibrs(void)
 
 static void __init noibpb(void)
 {
+	/*
+	 * Due to way how apply_forced_caps() works, we have to
+	 * explicitly clear the flag here from cas_set, otherwise it'll be
+	 * kept being put into the global mask.
+	 */
 	setup_clear_cpu_cap(X86_FEATURE_SPEC_CTRL);
+	clear_bit(X86_FEATURE_SPEC_CTRL, (unsigned long *)cpu_caps_set);
  	ibpb_state = 0;
 }
 
